Gradle & gitignore updates

HeinousGames opened the pull request.

glenjarvis commented on .idea/compiler.xml:
```diff
@@ -1,6 +0,0 @@
-<project version="4">
-  <component name="CompilerConfiguration">
-    <bytecodeTargetLevel target="11" />
-  </component>
```
@HeinousGames  Please help me understand the significance of `CompilerConfiguration`in an `idea/combiler.xml` file. Is this a local configuration of what compiler is to be used by a specific user?

glenjarvis commented on .idea/gradle.xml:
```diff
@@ -1,20 +0,0 @@
-        <option name="testRunner" value="PLATFORM" />
-        <option name="distributionType" value="DEFAULT_WRAPPED" />
-        <option name="externalProjectPath" value="$PROJECT_DIR$" />
-        <option name="modules">
```
@HeinousGames  Please help me understand the significance of this file. It appears to be a GradleMigration Settings for the project. Is this file a local file for individual customization?

glenjarvis commented on .idea/jarRepositories.xml:
```diff
@@ -1,25 +0,0 @@
-      <option name="id" value="jboss.community" />
-      <option name="name" value="JBoss Community repository" />
-      <option name="url" value="https://repository.jboss.org/nexus/content/repositories/public/" />
-    </remote-repository>
```
@HeinousGames  Why are these options not part of the project? How is the `.idea/` version of these files different than within the project.

For example, I would expect a list of JBOSS content to be a registered repository for the project `https://repository.jboss.org/nexus/content/repositories/public/`. Please help me understand the significance of the `.idea` forlder and why these should not be within the repo.

glenjarvis commented on .idea/modules/app/RothGar.app.iml:
```diff
@@ -1,157 +0,0 @@
-  <component name="FacetManager">
-    <facet type="android-gradle" name="Android-Gradle">
-      <configuration>
-        <option name="GRADLE_PROJECT_PATH" value=":app" />
```
What is the rational for the `app.iml` file? I've seen arguments both ways (to keep it and to remove it). Is this derived?

HeinousGames commented:
.idea files are generated and updated automatically by the IDE when you make changes. Rarely you will ever manually change the .idea files. You might make a dictionary.xml file so that your IDE doesn't throw spelling errors on custom named classes/objects, but that's the one instance where I've seen an Android project have manually made .idea files. The .idea files are a mix between user settings, compiler settings, and project settings. They will change if you:

- Update your IDE settings
- Update your IDE version
- Update the dependencies in either of your gradle files
- Update the jar dependencies your project uses
- Update the amount of modules you project has
- Change workspace settings
- Change VCS settings
- Change OS you are building on
- Change a million other things

Some files are tied to IDE functions. For instance, if I delete .idea/workspace.xml then Android Studio won't allow me to use the VCS UI tools anymore. If I close and reopen my project, it rebuilds the workspace.xml file and allows me access to the tools again. A lot of these files are generated based on project settings, so no matter what a lot of people will end up generating a lot of the same .idea files, while other files will be completely different (like ones pertaining to directory paths). There is debate about whether or not the .idea folder and some specific files should be shared or not. I believe in other projects it's more mandatory, but for us it shouldn't matter much. [JetBrains has a small article](https://intellij-support.jetbrains.com/hc/en-us/articles/206544839-How-to-manage-projects-under-Version-Control-Systems) about what .idea files are okay to push to the repo, and I believe some of our files fall within these categories. Nonetheless, for your small project, I don't think there's any harm in not committing any of them. Here's my reasoning:

- I deleted the entire .idea folder and committed the changes to my main fork. After deleting the project and re-cloning my fork without the .idea folder in the repo, the project generated its own .idea folder and was still able to compile. None of the people who clone your repo should be affected when building the src without changes, especially if they are already following your build requirements (Mac, Android Studio, etc.).
- When the new .idea folder is generated in Android Studio, the .gitignore file kicks in and actively shows me which files it is ignoring based on our rules. Every file is being ignored, and I did not make any changes to .gitignore regarding .idea, so your original rules are stating not to commit any of the files at all.
- Less files in the repo limits the time git takes to clone the repo.
- There's a chance of forcing a file onto somebody that affects their build, or a personal setting that wont work on their system. Chances are their IDE should regenerate these anyways though, but their first build might not realize and get stuck on a hardcoded file.

While I can't say what each tag in each .idea file means, I can say that the compiler.xml file will generate based on the compiler settings. The gradle.xml file will house gradle settings, as well as specific tags regarding the blocks you choose to implement in your gradle files. The jarRepositories.xml file will house settings surrounding the jar dependencies you add to your project, but not the gradle dependencies (old Android projects utilized most dependencies via jars before gradle was the default, so the option still exists alongside gradle dependencies). The JetBrains article I mentioned earlier is a little confusing regarding the .iml files. They say you can share the .iml files in your modules, but don't share the project's .iml file as the gradle system will generate it automatically. You're not using any extra modules, so your .iml file is one to ignore as well. Let me know if you still have any questions!

glenjarvis commented:
@HeinousGames  This explains a lot and is a good summary: `The .idea files are a mix between user settings, compiler settings, and project settings.`

This would explain why "some files are ignored and some aren't."

Ideally the compiler settings and project settings are checked-in and user settings are separate (not checked in).  And, for those who want that, they specifically ignore certain fails for user settings.

glenjarvis reviewed (approved):
Great feedback!

Ideally `.idea` would only be for user data (and not checked in) while the other portions (compiler settings, and project settings) stay within the project and are checked-in. This way we all build against a standard SDK, version, etc.

I use Git a lot in other projects to tell me what has changed so that I can catch drift (and thus why I would like some of those .idea files to be tracked). For example, if someone updated the SDK version, I'd want that to show in Git. The git status would show me that this was done. And, then I can decide either to update the SDK version to match on my side and/or talk to the other engineers and explain that I found drift.

However, as I know so little about Android Studio, gradle, etc. seeing those differences would not help inform me what should be changed -- I may see a lot of changes and not know what to do with them. For now, it is simplest to leave out all of the files.

glenjarvis commented:
Here is an additional follow-up note from Steve. I'm placing it here so we can find it when searching for it:

You mentioned keeping some .idea files for when you want to decide if you want to merge in certain changes like compiler or sdk settings. I'd like to point out that for most Android Studio projects you will end up seeing these changes in core code/gradle files, so the .idea files won't really matter. Here is a list of files that are in the repo, that when changed would alter .idea files on the user's local copy:
- /.build.gradle (this contains plugin versions, including the gradle plugin)
- /app/build.gradle (this contains the min, target, and compile sdk numbers, and dependency versions)
- /gradle/wrapper/gradle-wrapper.properties (literal gradle version that might need to change to run a newer gradle plugin version)
The .idea files that reflect these changes would show in the git change list, but it wouldn't be as obvious or make as much sense as the core variables they are derived from in the files listed above. Whenever you follow upgrade instructions for a major library release, you wont pay attention to the .idea files, you will end up changing objects in these files, so you may be putting extra work on yourself just to track .idea files for those settings.
